### DISCUSSION TITLE
Proposal to abstract away the API handlers into transport functions

### DISCUSSION THREAD

**Discussion by danhab99** [vercel/next.js#31136] [Ideas]
### Describe the feature you'd like to request

To prefix I have very little understanding of how webpack works. I'm simply going off of what I've seen next.js do.

One thing all webdevelopers have done in the past and still have to do with next.js is coordinate fetch requests between the webpage and a server side endpoint. This usually means having to write 2 kinds of code, the fetching code in the front, and an api endpoint in the back. This introduces boilerplate code and alot of room for mistakes to occur. It also makes it harder to typecheck what is going through a fetch request (if you're using typescript). 

What I'm proposing is to abstract that entire procedure into 1 single function that would allow a dev to write one code block to get frontend data into backend code from one concise codeblock. 

### Describe the solution you'd like

I'm thinking it would work similarly to how dynamically rendered pages can call their `getServerSideData` handler using the `next/router`. By calling `router.replace(router.asPath)` a page can refresh its entire state from the backend without having to reload, this is done by calling an endpoint generated at build time. This is why I think my idea is possible. 

In practice it would look something like this:

```ts
// ./pages/anypage.ts
import transport from "next/transporter";

/* What we see here is the `transporter` function. It will serve as some sort of black box for webpack 
 * to split the source into client and server side code. It will then return a fetcher function which would
 * perform the frontend query, doing it like this will allow the fetch function to be reused without redeclaring
 * the server side handler.
 */
const getUsers = transporter((body) => {
  // in here we're running server side code, that's how we have `Users` which is a mongoose model. `body`
  // has already been parsed into an object.
  const { skip, count } = body;
  return Users.find({})
    .skip(skip)
    .limit(count);
})

/* The fetcher function returns the promise of whatever type the transporter handler returns. Hopefully with
 * some typescript&webpack magic we can use this pattern to allow us full stack api layer type checking.
 */
getUsers({
  skip: 10,
  count: 10
}).then((users) => {
  if (Array.isArray(users)) {
    console.log(users)
  }
})
```

I imagine the type definition of the `transport` function looking something along the lines of this:

```ts
type transportHandler<B, R> = (body: B) => R | Promise<R>;

export type transport<B, R> = (handler: transportHandler<B, R>) => transportHandler<B, R>;
```

### Describe alternatives you've considered

The alternatives are obvious, we can just continue to create API endpoints. My idea for the transporter is more of a QOL solution, but its these little things that distinguish next.js as a full stack web framework.

**Comment by djsilcock** [comment]
You could check out [BlitzJS](https://blitzjs.com/) for an implementation of this (still in beta but looks like it has potential)
